mypyc/lib-rt/librt_internal.c

```diff
@@ -253,26 +258,47 @@ write_bool(PyObject *self, PyObject *const *args, size_t nargs, PyObject *kwname
 }
 
 /*
-str format: size followed by UTF-8 bytes
-    short strings (len <= 127): single byte for size as `(uint8_t)size << 1`
-    long strings: \x01 followed by size as Py_ssize_t
+str format: size as int (see below) followed by UTF-8 bytes
 */
 
+static inline CPyTagged
+_read_short_int(PyObject *data, uint8_t first) {
+    uint8_t second;
+    uint16_t two_more;
+    if ((first & TWO_BYTES_INT_BIT) == 0) {
+       // Note we use tagged ints since this function can return an error.
+       return ((Py_ssize_t)(first >> 1) + MIN_ONE_BYTE_INT) << 1;
+    }
+    if ((first & FOUR_BYTES_INT_BIT) == 0) {
+       _CHECK_READ(data, 1, CPY_INT_TAG)
+       second = _READ(data, uint8_t)
+       return ((((Py_ssize_t)second) << 6) + (Py_ssize_t)(first >> 2) + MIN_TWO_BYTES_INT) << 1;
+    }
+    // The caller is responsible to verify this is called only for short ints.
+    _CHECK_READ(data, 3, CPY_INT_TAG)
+    // TODO: check if compilers emit optimal code for these two reads, and tweak if needed.
+    second = _READ(data, uint8_t)
+    two_more = _READ(data, uint16_t)
```

Comment:
Since this code path will be quite rare, we could also read one byte at a time without any real performance impact. This would make this work the same on little and big endian systems.

Reply:
Btw this makes me think that right now this can produce some garbage on big-endian platforms without failing, which is not good. Also I found that there is `#if PY_BIG_ENDIAN` thing that we may use here.